File: internal/config/model_thinking_test.go
package config

import "testing"

func TestResolveModelThinkingUsesBuiltInPresets(t *testing.T) {
	tests := []struct {
		model        string
		requested    string
		wantLevel    string
		wantContains string
	}{
		{model: "openai/gpt-5.6", requested: "max", wantLevel: "max", wantContains: "xhigh"},
		{model: "anthropic/claude-opus-4.6", requested: "max", wantLevel: "max", wantContains: "high"},
		{model: "anthropic/claude-fable-5", requested: "xhigh", wantLevel: "xhigh", wantContains: "max"},
	}
	for _, test := range tests {
		profile := ResolveModelThinking(AppSettings{}, "tokenflux", test.model, test.requested)
		if !profile.Enabled || profile.Level != test.wantLevel || !containsThinkingLevel(profile.Levels, test.wantContains) {
			t.Fatalf("ResolveModelThinking(%q) = %#v", test.model, profile)
		}
	}
}

func TestResolveModelThinkingRequiresManualOptInForOtherModels(t *testing.T) {
	settings := AppSettings{
		ModelThinking: map[string]map[string]ModelThinkingConfig{
			"tokenflux": {
				"x-ai/grok-4.6": {
					Enabled:      true,
					Levels:       []string{"low", "high"},
					DefaultLevel: "high",
				},
			},
		},
	}
	if profile := ResolveModelThinking(AppSettings{}, "tokenflux", "x-ai/grok-4.6", "high"); profile.Enabled {
		t.Fatalf("unexpected built-in Grok thinking profile: %#v", profile)
	}
	profile := ResolveModelThinking(settings, "tokenflux", "x-ai/grok-4.6", "medium")
	if !profile.Enabled || profile.Level != "high" {
		t.Fatalf("manual Grok thinking profile = %#v", profile)
	}
}

func TestNormalizeModelThinkingDropsUnsupportedLevelsAndCanDisablePreset(t *testing.T) {
	settings := AppSettings{
		Providers: map[string]ProviderConfig{},
		ModelThinking: map[string]map[string]ModelThinkingConfig{
			"tokenflux": {
				"openai/gpt-5.6": {
					Enabled:      true,
					Levels:       []string{"medium", "ultra", "max", "medium"},
					DefaultLevel: "ultra",
				},
				"openai/gpt-5.5": {Enabled: false},
			},
		},
	}
	normalized := withDefaults(settings)
	configured := normalized.ModelThinking["tokenflux"]["openai/gpt-5.6"]
	if got := len(configured.Levels); got != 2 {
		t.Fatalf("normalized levels = %#v", configured.Levels)
	}
	if configured.Levels[0] != "medium" || configured.Levels[1] != "max" || configured.DefaultLevel != "medium" {
		t.Fatalf("normalized config = %#v", configured)
	}
	if profile := ResolveModelThinking(normalized, "tokenflux", "openai/gpt-5.5", "high"); profile.Enabled {
		t.Fatalf("manual disable did not override preset: %#v", profile)
	}
}

func TestNormalizeModelThinkingRemembersConfiguredLevelsWhileDisabled(t *testing.T) {
	configured := normalizeModelThinkingConfig(ModelThinkingConfig{
		Enabled:      false,
		Levels:       []string{"low", "high", "max"},
		DefaultLevel: "high",
	})
	if len(configured.Levels) != 3 || configured.DefaultLevel != "high" {
		t.Fatalf("disabled thinking config lost its choices: %#v", configured)
	}
}

File: app/src/lib/modelThinking.test.ts
import { describe, expect, it } from 'vitest'
import {
  effectiveModelThinkingLevel,
  MODEL_THINKING_LEVEL_LABELS,
  normalizeModelThinkingConfig,
  resolveModelThinking,
} from '@/lib/modelThinking'
import type { AppSettings } from '@/types'

const settings = (modelThinking?: AppSettings['model_thinking']) => ({
  model_thinking: modelThinking,
})

describe('model thinking profiles', () => {
  it('uses the canonical English level names as user-visible labels', () => {
    expect(MODEL_THINKING_LEVEL_LABELS).toEqual({
      off: 'off',
      minimal: 'minimal',
      low: 'low',
      medium: 'medium',
      high: 'high',
      xhigh: 'xhigh',
      max: 'max',
    })
  })

  it('presets reasoning-capable families from models.dev effort lists', () => {
    expect(resolveModelThinking(settings(), 'tokenflux', 'openai/gpt-5.6')).toMatchObject({
      enabled: true,
      defaultLevel: 'medium',
      levels: ['off', 'low', 'medium', 'high', 'xhigh', 'max'],
    })
    expect(resolveModelThinking(settings(), 'tokenflux', 'anthropic/claude-fable-5')).toMatchObject({
      enabled: true,
      defaultLevel: 'high',
      levels: ['low', 'medium', 'high', 'xhigh', 'max'],
    })
    expect(resolveModelThinking(settings(), 'tokenflux', 'deepseek-flash')).toMatchObject({
      enabled: true,
      defaultLevel: 'high',
      levels: ['low', 'high', 'max'],
    })
    expect(resolveModelThinking(settings(), 'tokenflux', 'deepseek-v4-pro')).toMatchObject({
      enabled: true,
      defaultLevel: 'high',
      levels: ['low', 'high', 'max'],
    })
    expect(resolveModelThinking(settings(), 'tokenflux', 'x-ai/grok-4.7')).toMatchObject({
      enabled: true,
      defaultLevel: 'medium',
      levels: ['low', 'medium', 'high', 'xhigh'],
    })
    expect(resolveModelThinking(settings(), 'tokenflux', 'x-ai/grok-4.6')).toMatchObject({
      enabled: true,
      defaultLevel: 'medium',
      levels: ['low', 'medium', 'high', 'xhigh'],
    })
  })

  it('requires an explicit override for models outside the preset families', () => {
    expect(resolveModelThinking(settings(), 'tokenflux', 'vendor/unknown-chat').enabled).toBe(false)
    const configured = settings({
      tokenflux: {
        'vendor/unknown-chat': {
          enabled: true,
          levels: ['low', 'high'],
          default_level: 'high',
        },
      },
    })
    const profile = resolveModelThinking(configured, 'tokenflux', 'vendor/unknown-chat')
    expect(profile.source).toBe('manual')
    expect(effectiveModelThinkingLevel(profile, 'medium')).toBe('high')
  })

  it('filters unsupported labels including ultra', () => {
    expect(normalizeModelThinkingConfig({
      enabled: true,
      levels: ['medium', 'ultra', 'max'] as never[],
      default_level: 'ultra' as never,
    })).toEqual({
      enabled: true,
      levels: ['medium', 'max'],
      default_level: 'medium',
    })
    expect(normalizeModelThinkingConfig({
      enabled: false,
      levels: ['low', 'high', 'max'],
      default_level: 'high',
    })).toEqual({
      enabled: false,
      levels: ['low', 'high', 'max'],
      default_level: 'high',
    })
  })
})
